python/ros_interface.py

```diff
@@ -0,0 +1,110 @@
+        self.past_gps_lat = self.gps_lat
+        self.past_gps_lon = self.gps_lon
+        self.past_gps_headingDegrees = self.gps_headingDegrees
+        self.past_gps_speedKmph = self.gps_speedKmph
+        self.past_wind_speedKmph = self.wind_speedKmph
+        self.past_wind_direction = self.wind_direction
+
+        # GPS sensors
+        self.gps_lat = self.getTrustedAndAvg(self.past_gps_lat,
+                                             [self.data.gps_0_latitude, self.data.gps_1_latitude])
+        self.gps_lon = self.getTrustedAndAvg(self.past_gps_lon,
+                                             [self.data.gps_0_longitude, self.data.gps_1_longitude])
+        self.gps_headingDegrees = self.getTrustedAndAvg(self.past_gps_headingDegrees,
+                                                        [self.convertDegrees(self.data.gps_0_true_heading),
+                                                         self.convertDegrees(self.data.gps_1_true_heading)])
+        self.gps_speedKmph = self.getTrustedAndAvg(self.past_gps_speedKmph,
+                                                   [self.data.gps_0_groundspeed * KNOTS_TO_KMPH,
+                                                    self.data.gps_1_groundspeed * KNOTS_TO_KMPH])
+        # Wind sensors
+        self.wind_speedKmph = self.getTrustedAndAvg(self.past_wind_speedKmph,
+                                                    [self.data.wind_sensor_0_speed * KNOTS_TO_KMPH,
+                                                     self.data.wind_sensor_1_speed * KNOTS_TO_KMPH,
+                                                     self.data.wind_sensor_2_speed * KNOTS_TO_KMPH])
+        self.wind_direction = self.getTrustedAndAvg(self.past_wind_direction,
+                                                    [self.data.wind_sensor_0_direction,
+                                                     self.data.wind_sensor_1_direction,
+                                                     self.data.wind_sensor_2_direction])
+
+    # (0 is north, 90 is east, etc - cw - degrees) -> (0 is east, 90 is north, etc - ccw - degrees)
+    def convertDegrees(self, degree):
+        converted = -1 * degree + 90
+        return converted if converted >= 0 else converted + 360
+
+    def getTrustedAndAvg(self, pastValue, vals):
+        err_free_vals = self.getTrusted(pastValue, vals)
+        if not err_free_vals:
+            rospy.logwarn("Values don't match with pastValue, or pastValue is 0, defaulting to first sensor's data")
+            return vals[0]
+        else:
+            return float(sum(err_free_vals)) / len(err_free_vals)
+
+    # assumes sensor readings are consistent, first sensor initial reading is accurate
+    # could modify past value to be the average of the past few outputs (store in list)
+    def getTrusted(self, pastValue, vals):
+        errorFreeVals = []
+        for value in vals:
+            if abs(float((pastValue - value)) / pastValue) < pastValue * ERROR:
```

Comment:
I think you left in the `/ pastValue` part by accident. I see what you were doing to avoid division by 0 and still comparing the same, but it's not quite as clear. Maybe like

```
if pastValue != 0 and abs(float(pastValue - value) / pastValue) < ERROR:
```